francois.role@parisdescartes.fr

### Let $X$ be a N * M document-term matrix

In [5]:
import numpy as np
import scipy.linalg as linalg

In [6]:
X=np.array([ [1., 1., 1., 0., 0.] ,
             [2., 2., 2., 0. ,0.],
             [1., 1. ,1., 0., 0.] ,
             [5., 5. ,5. ,0., 0.] ,
             [0., 0. ,0., 2., 2.] ,
             [0., 0., 0. ,3. ,3.] ,
             [0., 0., 0. ,1.,1.]
              ])
# cols = car vehicle truck cat dog

In [7]:
print(np.linalg.matrix_rank(X))

2


### Compute the eigenvectors of $XX^t$ and use them as columns of the $U$ matrix (documents)

In [8]:
XXt=X.dot(X.T)
eigVals, eigVects= linalg.eig(XXt)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
topEigVects=eigVects[:, eigValsIndices]

print (topEigVects)
print
print (np.sqrt(eigVals[eigValsIndices]))  # singular values
U=topEigVects
S=np.diag(np.sqrt(eigVals[eigValsIndices]))


[[ 0.1796053   0.        ]
 [ 0.3592106   0.        ]
 [ 0.1796053   0.        ]
 [ 0.89802651  0.        ]
 [ 0.         -0.53452248]
 [ 0.         -0.80178373]
 [ 0.         -0.26726124]]
[ 9.64365076+0.j  5.29150262+0.j]


### Compute the eigenvectors of $X^tX$ and use them as columns of the $V$ matrix (terms)

In [48]:
XtX=X.T.dot(X)
eigVals, eigVects= linalg.eig(XtX)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
topEigVects=eigVects[:, eigValsIndices]

print (topEigVects)
print
print (np.sqrt(eigVals[eigValsIndices]) ) # singular values
V=topEigVects

[[ 0.57735027  0.        ]
 [ 0.57735027  0.        ]
 [ 0.57735027  0.        ]
 [ 0.          0.70710678]
 [ 0.          0.70710678]]
[ 9.64365076+0.j  5.29150262+0.j]


### Reconstruct the matrix. We can do it exactly. Why?

In [10]:
print(U.dot(S).dot(V.T))

print()

print(X)
#U.shape
#S.shape
#V.T.shape

[[ 1.+0.j  1.+0.j  1.+0.j  0.+0.j  0.+0.j]
 [ 2.+0.j  2.+0.j  2.+0.j  0.+0.j  0.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  0.+0.j  0.+0.j]
 [ 5.+0.j  5.+0.j  5.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j -2.+0.j -2.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j -3.+0.j -3.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j -1.+0.j -1.+0.j]]

[[ 1.  1.  1.  0.  0.]
 [ 2.  2.  2.  0.  0.]
 [ 1.  1.  1.  0.  0.]
 [ 5.  5.  5.  0.  0.]
 [ 0.  0.  0.  2.  2.]
 [ 0.  0.  0.  3.  3.]
 [ 0.  0.  0.  1.  1.]]


### Check the distance between the original and reconstructed matrix

In [11]:
reconstructed=np.abs(U.dot(S).dot(V.T))

np.sqrt(np.sum ( (X - reconstructed)**2  ) )  # Frobenius

1.8709810974240868e-15

### The first document is represented in the latent space by the first row of matrix $U$ :
   
[ 0.1796053 ,   0.  ]

### The relation between the original representation of the documents and their representation in the latent semantic space can be derived as follows:
$$
X=U \Sigma V^T   \Leftrightarrow X V {\Sigma}^{-1} = U
$$

Hence the "conceptual" representation of the documents (the $U$ matrix) can be obtained by projecting the original representation of the documents into the term latent space $V$.

In [19]:
S_inv=np.mat(S).I
np.abs(X.dot(V).dot(S_inv))

matrix([[ 0.1796053 ,  0.        ],
        [ 0.3592106 ,  0.        ],
        [ 0.1796053 ,  0.        ],
        [ 0.89802651,  0.        ],
        [ 0.        ,  0.53452248],
        [ 0.        ,  0.80178373],
        [ 0.        ,  0.26726124]])

In [ ]:
U

In the same spirit, we can derive $V$ from $U$ and $X$:
$$
X=U \Sigma V^T   \Leftrightarrow V= X^{-1}U{\Sigma}
$$ 

### Verify it

In [21]:
X_inv=np.mat(X).I
np.abs(X_inv.dot(U).dot(S))

matrix([[  5.77350269e-01,   0.00000000e+00],
        [  5.77350269e-01,   0.00000000e+00],
        [  5.77350269e-01,   0.00000000e+00],
        [  6.66333430e-34,   7.07106781e-01],
        [  7.49540088e-34,   7.07106781e-01]])

### Application to Information Retrieval

This leads to useful applications. For example, in the Information Retrieval area, suppose we are given a query $q$. We can treat $q$ as a short document and project it into the latent space.

In [53]:
X=np.mat([   [1., 1., 0., 0., 0.] ,
             [0., 0., 2., 0. ,0.],
             [1., 0. ,1., 0., 0.] ,
             [0., 0., 5. ,0., 0.] ,
             [0., 0. ,0., 2., 2.] ,
             [0., 0., 0. ,3. ,3.] ,
             [0., 0., 0. ,1.,1.]
              ])
# cols = car vehicle truck cat dog

q=np.mat([1., 1., 0., 0., 0.]).T

Using a standard query model, we would miss the second and fourth document altough they are related to the automotive sector.

In [54]:
X * q

matrix([[ 2.],
        [ 0.],
        [ 1.],
        [ 0.],
        [ 0.],
        [ 0.],
        [ 0.]])

The solution is to project $q$ into the latent space. If $q$ is represented as a row vector, the projected query $\hat{q}$ will be obtained as:

$$
\hat{q}= q V {\Sigma}^{-1}
$$

In [65]:
XXt=X.dot(X.T)
eigVals, eigVects= linalg.eig(XXt)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
U=eigVects[:, eigValsIndices]

S=np.mat(np.diag(np.sqrt(eigVals[eigValsIndices])))

XtX=X.T.dot(X)
eigVals, eigVects= linalg.eig(XtX)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
V=eigVects[:, eigValsIndices]

print(U)
print("====")
print(S)
print("====")
print(V)


S_inv=S.I
q_hat= q.T * V * S_inv  # q.T = a row like a doc in a doc-term matrix

q_hat= np.abs(np.array(q_hat).flatten() )
U.dot(q_hat)


[[ 0.00673646  0.        ]
 [ 0.36469846  0.        ]
 [ 0.1888614   0.        ]
 [ 0.91174614  0.        ]
 [ 0.         -0.53452248]
 [ 0.         -0.80178373]
 [ 0.         -0.26726124]]
====
[[ 5.48048471+0.j  0.00000000+0.j]
 [ 0.00000000+0.j  5.29150262+0.j]]
====
[[-0.03568988  0.        ]
 [-0.00122917  0.        ]
 [-0.99936216  0.        ]
 [ 0.          0.70710678]
 [ 0.          0.70710678]]
<class 'numpy.ndarray'>


We were able to match documents (second and fourth) that were about automotive area but had no terms in common with our query!


********************************************************************************************

### If we project the conceptual representation of the 4th doc onto V, we get back the original representation of the 4th doc :
$$
X_{j.}= \hat{V} \hat{\Sigma } {\hat{U}_{j.}}^T
$$

In [64]:
V.dot(S).dot(U[3,:].T)

matrix([[-0.17833560+0.j, -0.00614194+0.j, -4.99362361+0.j,  0.00000000+0.j,
          0.00000000+0.j]])

### We now know how to project a document into the k-dimensional latent space:
$$
{\hat{U}_{j.}}^T= ({\hat{\Sigma }}^{-1} {\hat{V}}^T) X_{j.}
$$

### Verify that we get the same result as that obtained from Scipy

In [ ]:
from scipy import linalg
U , S , V= linalg.svd(X)
print ("U")
print( np.c_[  np.mat(U[:,0]).T , np.mat(U[:,1]).T ])
print 
print ("S")
print (S)
print
print ("V^t")
print ( np.r_[  np.mat(V[0,:]) , np.mat(V[1,:]) ])
U

In [ ]:
S=np.mat(S)
S.I.shape
V.T
U

In [ ]:
# To do Frobenius norm

### Project a document

Project the documents onto the eigenvectors of $X^tX$ ($V$, "term to concept" matrix").
Document $4$ has an "affinity" of $8.6$ with the "first concept".

In [ ]:
eigVals, eigVects= linalg.eig(XtX)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
topEigVects=eigVects[:, eigValsIndices]

print (X.dot(topEigVects))

Project the terms onto the eigenvectors of $XX^t$ ($U$, "document to concept" matrix)

In [ ]:
eigVals, eigVects= linalg.eig(XXt)
eigValsIndices=eigVals.argsort()
eigValsIndices=eigValsIndices[:-3:-1] 
topEigVects=eigVects[:, eigValsIndices]

print( X.T.dot(topEigVects))

In [13]:
# Exemple Master 2 ///////////////////////////////////////////////////////////////

A=np.array( [ [0,1,0,1,1,0,1],
[0,1,1,0,0,0,0],
[0,0,0,0,0,1,1],
[0,0,0,1,0,0,0],
[0,1,1,0,0,0,0],
[1,0,0,1,0,0,0],
[0,0,0,0,1,1,0],
[0,0,1,1,0,0,0],
[1,0,0,1,0,0,0 ] ], dtype=float)

A_col_norms=np.linalg.norm(A, axis=0)
A_col_norms_matrix = np.tile(A_col_norms, (9,1))
A_col_normalized= A * (1. / A_col_norms_matrix)

q = np.array([1, 0, 0, 1 ,0, 0, 0, 0, 0])
q_norm=np.linalg.norm(q)
q_normalized= q / q_norm

print np.linalg.norm(A_col_normalized, axis=0)

simScores= q_normalized.dot(A_col_normalized)
print simScores



SyntaxError: invalid syntax (<ipython-input-13-f0dbf86bc46a>, line 21)

In [14]:
# /////////  FULL SVD   //////

a = A
U, s, V = np.linalg.svd(a, full_matrices=True)
print  U.shape, V.shape, s.shape
S = np.zeros((9, 7), dtype=float)
S[:7, :7] = np.diag(s)
print
print S
print
print a
print
print np.dot(U, np.dot(S, V))

# ////////////  Frobenius Norm  ///////

print np.linalg.norm(A -   np.dot(U, np.dot(S, V))  )


SyntaxError: invalid syntax (<ipython-input-14-2f8461f20076>, line 5)

In [15]:
# /////////////  RANK 4 approximation  //////////

U, s, V = np.linalg.svd(A, full_matrices=True)

print U.shape

print s

print V.shape

U_4=U[:,:7]
V_4=V[:7,:]
S_4=np.diag(s[:7])

print  U_4.shape, V_4.shape, S_4.shape

print

print A
print
A_4=np.dot(U_4, np.dot(S_4, V_4))
print A_4 # or  U_4.dot(s_4).dot(V_4)

# ////////////  Frobenius Norm  ///////

print "Frob. norm" , np.linalg.norm(A - A_4)

# ///////  Normalize the reduced A_4   /////////

print np.linalg.norm(A_4 , axis=0)

A4_col_norms=np.linalg.norm(A_4, axis=0)
A4_col_norms_matrix = np.tile(A4_col_norms, (9,1))
A4_col_normalized= A_4 * (1. / A4_col_norms_matrix)

print np.linalg.norm(A4_col_normalized , axis=0)

# /////  Query the reduced matrix //////

q = np.array([1, 0, 0, 1 ,0, 0, 0, 0, 0])
q_norm=np.linalg.norm(q)
q_normalized= q / q_norm

simScores= q_normalized.dot(A4_col_normalized)
print simScores





SyntaxError: invalid syntax (<ipython-input-15-d7f7db9f243c>, line 5)